### PR TITLE
Adding ignore_index to segmentation mean_iou

### LINKED ISSUE #2747: Segmentation IOU compute Ignore some tagged values that don't need to be recorded (such as 255)
## 🚀 Feature

when we compute IOU
```python
import torch

_ = torch.manual_seed(0)
from torchmetrics.segmentation import MeanIoU

miou = MeanIoU(num_classes=3)
preds = torch.randint(0, 2, (5,))
target = torch.as_tensor((0, 1, 2, 0, 255))  # An index of 255 is a tag to be ignored.
miou(preds, target)
>>> This will result in an error
```

### Motivation

When I generate the sample pairs, the opposite mask (assuming 3 classes), but not all pixels in the entire mask should be classified into a particular class, so I set these pixels to 255. The pixel is then ignored in the loss calculation using `torch.nn.CrossEntropyLoss(ignore_index=255)`. value. However, the IOU calculation does not have this feature, which leads to errors in the IOU calculation, so I wondered if it could be made to support the ignore_index parameter as well, to ignore certain pixels.

### Pitch

```python
import torch
_ = torch.manual_seed(0)
from torchmetrics.segmentation import MeanIoU
miou = MeanIoU(num_classes=3, ignore_index=255)  # support ignore_index param to ignore index 255
preds = torch.randint(0, 2, (5,))
target = torch.as_tensor((0, 1, 2, 0, 255))  # An index of 255 is a tag to be ignored.
miou(preds, target)
```

### Alternatives

https://github.com/Lightning-AI/torchmetrics/blob/62d9d32280ba365f1c2c14e0bd8a5adc959a1a6e/src/torchmetrics/functional/segmentation/mean_iou.py#L42

https://github.com/Lightning-AI/torchmetrics/blob/62d9d32280ba365f1c2c14e0bd8a5adc959a1a6e/src/torchmetrics/functional/segmentation/mean_iou.py#L52-L55

```python
def _mean_iou_update(
    preds: Tensor,
    target: Tensor,
    num_classes: int,
    include_background: bool = False,
    input_format: Literal["one-hot", "index"] = "one-hot",
    ignore_index=255

) -> Tuple[Tensor, Tensor]:
    ...

    if input_format == "index":
        preds = torch.nn.functional.one_hot(preds, num_classes=num_classes).movedim(-1, 1)
        mask = torch.where(target == ignore_index)  #  Add removal of ignored labels
        target[mask] = 0
        target = torch.nn.functional.one_hot(target, num_classes=num_classes)
        target[mask] = 0  # set ont-hot to zero-hot from ignored labels
        target = target.movedim(-1, 1)
  ...
```

### PR DESCRIPTION
**Pull request by VijayVignesh1**
## What does this PR do?

Fixes #2747 

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>
Added ignore_index to segmentation mean_iou class and function. Users can now add an index in the target which can be ignored (like 255). Added testcases to verify the same.
Note: This is only for the case where input_format = "index"

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Changed
 
 - Defaulting Dice score `average="macro"` ([#3042](https://github.com/Lightning-AI/torchmetrics/pull/3042))
+- Added `ignore_index` to Segmentation IOC metric ([#2747](https://github.com/Lightning-AI/torchmetrics/issues/2747))
```
Comment (Borda) [inline CHANGELOG.md]:
```suggestion


- Added `ignore_index` to Segmentation IOC metric ([#2747](https://github.com/Lightning-AI/torchmetrics/issues/2747))
```

**Review by Borda — APPROVED** [review]

File: tests/unittests/segmentation/test_mean_iou.py [hunk 2]
```diff
@@ -83,11 +86,14 @@ def _reference_mean_iou(
 class TestMeanIoU(MetricTester):
     """Test class for `MeanIoU` metric."""
```
[the remark this answers is not in the document]
Comment (VijayVignesh1) [reply tests/unittests/segmentation/test_mean_iou.py:88]: No additional comment needed here since the rationale is self-understood.

File: tests/unittests/segmentation/test_mean_iou.py [hunk 3]
```diff
@@ -197,3 +212,27 @@ def test_mean_iou_perfect_prediction():
...
+    ])
+    metric.update(preds, target)
+    miou_per_class = metric.compute()
+    miou_func = mean_iou(preds, target, num_classes=3, per_class=True, input_format="index", ignore_index=255).mean(
+        dim=0
+    )  # reduce over batch dim
+    expected_ious = [0.6667, 0.3333, 0.0]
+    for idx, (iou, iou_func) in enumerate(zip(miou_per_class, miou_func)):
+        assert torch.allclose(iou, iou_func, atol=1e-4)
+        assert iou == pytest.approx(expected_ious[idx], rel=1e-3)
```
[the remark this answers is not in the document]
Comment (VijayVignesh1) [reply tests/unittests/segmentation/test_mean_iou.py:238]: No additional comment needed here since the rationale is self-understood.

File: src/torchmetrics/functional/segmentation/mean_iou.py [hunk 4]
```diff
@@ -72,8 +72,13 @@ def _mean_iou_update(
     num_classes: Optional[int] = None,
     include_background: bool = False,
     input_format: Literal["one-hot", "index", "mixed"] = "one-hot",
+    ignore_index: Optional[int] = None,
 ) -> tuple[Tensor, Tensor]:
     """Update the intersection and union counts for the mean IoU computation."""
+    if ignore_index is not None and input_format == "index":
+        idx = target == ignore_index
```
Comment (justusschock) [inline src/torchmetrics/functional/segmentation/mean_iou.py]: can we support that also in onehot by basically just removing the row in that dimension?

File: src/torchmetrics/functional/segmentation/mean_iou.py [hunk 5]
```diff
@@ -72,8 +72,13 @@ def _mean_iou_update(
...
+        target, preds = target[~idx], preds[~idx]
```
Comment (justusschock) [inline src/torchmetrics/functional/segmentation/mean_iou.py]: let's directly compute idx = target != ignore_index. Otherwise ~idx will create an additional intermediate tensor leading to additional memory allocation

File: src/torchmetrics/functional/segmentation/mean_iou.py (hunk 4 above)
Comment (VijayVignesh1) [reply src/torchmetrics/functional/segmentation/mean_iou.py]: I could do that. However, in most cases the ignore_index value is 255 which will blow up the one-hot tensor, right? So, for one-hot tensors can we assume that the user would handle the masking?
